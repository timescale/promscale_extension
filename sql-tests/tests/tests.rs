use insta::assert_snapshot;
use std::{env, sync::{Mutex, Once}, mem::MaybeUninit};
use test_common::*;
use test_generator::test_resources;

const TESTDATA: &'static str = concat!(env!("CARGO_MANIFEST_DIR"), "/testdata");

fn init_lock() -> &'static Mutex<()> {
    static mut SINGLETON: MaybeUninit<Mutex<()>> = MaybeUninit::uninit();
    static ONCE: Once = Once::new();

    unsafe {
        ONCE.call_once(|| {
            let singleton = Mutex::new(());
            SINGLETON.write(singleton);
        });

        SINGLETON.assume_init_ref()
    }
}

#[test_resources("testdata/*.sql")]
fn sql_tests(resource: &str) {
    let pg_blueprint = PostgresContainerBlueprint::new().with_testdata(TESTDATA);
    let test_pg_instance = new_test_instance_from_env(&pg_blueprint);
    let mut init_conn = test_pg_instance.connect();
    // This lock prevents multiple callers from running 
    // CREATE EXTENSION simultaneously. In case of
    // promscale_extension concurrent CREATE EXTENSION causes
    // issues when migration incremental/003-users.sql executes.
    {
        let _init_lock = init_lock().lock().unwrap();
        init_conn
            .simple_query("CREATE EXTENSION promscale;")
            .expect("Unable to create extension promscale.");
    }

    let query_result = test_pg_instance.exec_sql_script(resource);
    assert_snapshot!(resource, query_result);
}
