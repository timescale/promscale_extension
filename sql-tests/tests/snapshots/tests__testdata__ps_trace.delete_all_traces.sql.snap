---
source: sql-tests/tests/tests.rs
expression: query_result
---
 plan  
-------
 1..16
(1 row)

 put_operation 
---------------
             1
(1 row)

 put_tag_key 
-------------
        1001
(1 row)

 put_tag 
---------
       2
(1 row)

                  is                   
---------------------------------------
 ok 1 - _ps_trace.schema_url has 1 row
(1 row)

                       is                       
------------------------------------------------
 ok 2 - _ps_trace.instrumentation_lib has 1 row
(1 row)

                  is                  
--------------------------------------
 ok 3 - _ps_trace.operation has 1 row
(1 row)

                 is                 
------------------------------------
 ok 4 - _ps_trace.tag_key has 1 row
(1 row)

               is                
---------------------------------
 ok 5 - _ps_trace.tag has 2 rows
(1 row)

               is                
---------------------------------
 ok 6 - _ps_trace.span has 1 row
(1 row)

               is                
---------------------------------
 ok 7 - _ps_trace.link has 1 row
(1 row)

                is                
----------------------------------
 ok 8 - _ps_trace.event has 1 row
(1 row)

psql:testdata/ps_trace.delete_all_traces.sql:73: NOTICE:  truncate cascades to table "instrumentation_lib"
 delete_all_traces 
-------------------
 
(1 row)

                   is                   
----------------------------------------
 ok 9 - _ps_trace.schema_url has 0 rows
(1 row)

                        is                        
--------------------------------------------------
 ok 10 - _ps_trace.instrumentation_lib has 0 rows
(1 row)

                   is                   
----------------------------------------
 ok 11 - _ps_trace.operation has 0 rows
(1 row)

                  is                  
--------------------------------------
 ok 12 - _ps_trace.tag_key has 0 rows
(1 row)

                is                
----------------------------------
 ok 13 - _ps_trace.tag has 0 rows
(1 row)

                is                 
-----------------------------------
 ok 14 - _ps_trace.span has 0 rows
(1 row)

                is                 
-----------------------------------
 ok 15 - _ps_trace.link has 0 rows
(1 row)

                 is                 
------------------------------------
 ok 16 - _ps_trace.event has 0 rows
(1 row)

 finish 
--------
(0 rows)


